pilot/pkg/dns/dns.go

```diff
@@ -0,0 +1,344 @@
+	h := &LocalDNSServer{
+		downstreamMux:  dns.NewServeMux(),
+		proxyNamespace: proxyNamespace,
+	}
+
+	// proxyDomain could contain the namespace making it redundant.
+	// we just need the .svc.cluster.local piece
+	parts := strings.Split(proxyDomain, ".")
+	if len(parts) > 0 {
+		if parts[0] == proxyNamespace {
+			parts = parts[1:]
+		}
+		h.proxyDomainParts = parts
+		h.proxyDomain = strings.Join(parts, ".")
+	}
+
+	h.downstreamMux.Handle(".", h)
+
+	h.upstreamClient = &dns.Client{
+		Net:         "udp",
+		DialTimeout: 3 * time.Second,
+		ReadTimeout: 500 * time.Millisecond,
+	}
+
+	// We will use the local resolv.conf for resolving unknown names.
+	dnsConfig, err := dns.ClientConfigFromFile("/etc/resolv.conf")
+	if err != nil {
+		log.Warnf("failed to load /etc/resolv.conf: %v", err)
+		return nil, err
+	}
+
+	// Unlike traditional DNS resolvers, we do not need to append the search
+	// namespace to a given query and try to resolve it. This is because the
+	// agent acts as a DNS interceptor for DNS queries made by the application.
+	// The application's resolver is already sending us DNS queries, one for each
+	// of the DNS search namespaces. We simply need to check the existence of this
+	// name in our local nametable. If not, we will forward the query to the
+	// upstream resolvers as is.
+	//
+	// NOTE: There is still a NxN issue here as the app will cycle through N
+	// nameservers for each potential FQDN string. The agent will also cycle through
+	// NxN unfortunately as it cannot infer the orig dst IP from the incoming packet.
+	if dnsConfig != nil {
+		if len(dnsConfig.Servers) > 0 {
+			for _, s := range dnsConfig.Servers {
+				h.resolvConfServers = append(h.resolvConfServers, s+":53")
+			}
+		}
+		h.searchNamespaces = dnsConfig.Search
+	}
+	return h, nil
+}
+
+// StartDNS starts the DNS-over-UDP downstreamServer.
+func (h *LocalDNSServer) StartDNS() {
+	var err error
+	h.downstreamServer = &dns.Server{Handler: h.downstreamMux}
+	h.downstreamServer.PacketConn, err = net.ListenPacket("udp", ":15053")
+	if err != nil {
+		log.Warna("Failed to start DNS downstreamServer", err)
+	}
+
+	log.Infoa("Started DNS downstreamServer at 0.0.0.0:15053")
+	go func() {
+		err := h.downstreamServer.ActivateAndServe()
+		if err != nil {
+			log.Errora("Failed to activate DNS downstreamServer ", err)
+		}
+	}()
+}
+
+func (h *LocalDNSServer) UpdateLookupTable(nt *nds.NameTable) {
+	lookupTable := &LookupTable{
+		name4: map[string][]dns.RR{},
+		name6: map[string][]dns.RR{},
+		cname: map[string][]dns.RR{},
+	}
+	for host, ni := range nt.Table {
+		// Given a host
+		// if its a non-k8s host, store the host+. as the key with the pre-computed DNS RR records
+		// if its a k8s host, store all variants (i.e. shortname+., shortname+namespace+., fqdn+., etc.)
+		// shortname+. is only for hosts in current namespace
+		var altHosts []string
+		if ni.Registry == "Kubernetes" {
```

Comment:
are multicluster in another cluster still `"Kubernetes"`?

Reply:
I would hope so.. because whatever istiod reads from k8s, it marks as kubernetes. If its replicated control planes, then there is no way a local istiod can know about it unless one creates a service entry (and this is not what we advertise)